libvast/src/system/index.cpp

```diff
@@ -440,10 +440,15 @@ index(index_actor::stateful_pointer<index_state> self,
     // Persist active partition asynchronously.
     auto part_dir = dir / to_string(id);
     VAST_DEBUG(self, "persists active partition to", part_dir);
-    self->request(actor, caf::infinite, atom::persist_v, part_dir, self)
+    self->request(actor, caf::infinite, atom::persist_v, part_dir /*, self*/)
       .then(
-        [=](atom::ok) {
+        [=](atom::ok, std::shared_ptr<partition_synopsis>& ps) {
           VAST_DEBUG(self, "successfully persisted partition", id);
+          // Semantically ps is a unique_ptr, and the partition releases its
+          // copy before sending. We use shared_ptr for the transport because
+          // CAF message types must be copy-constructible.
+          VAST_ASSERT(ps.use_count() <= 1);
```

Comment:
Shouldn't this check whether the shared_ptr is unique exactly? If the use count is 0, then you dereference a nullptr directly below.